### PR TITLE
Requires platform flag only when ambiguous

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This allows us to peek at images even if the platform they have listed is wrong or irrelevant.

Ex.
```bash
./car -tvvf hydai/wasm-wasi-example:latest
https://index.docker.io/v2/hydai/wasm-wasi-example/manifests/latest platform=linux/amd64 totalLayerSize: 343437
https://index.docker.io/v2/hydai/wasm-wasi-example/blobs/sha256:6ed9d2593eddde644c7037a80b63a5e37fd928bdd014f40341515833dc3ace6d size=343437
CreatedBy: /bin/sh -c #(nop) ADD file:603e543aa882bc3cb282c8885d8c6ab7acd0c27f7f457d818e7943cf325a8bb5 in . 
-rw-rw-r--	1556954	Jun 25 21:07:25	wasi_example_main.wasm
```

### CODE REVIEW

File: internal/registry/registry.go [hunk 1]
```diff
@@ -49,7 +50,11 @@ func New(ctx context.Context, host, path string) internal.Registry {
 		path = pathutil.Join("library", path)
 	}
 	transport := httpClientTransport(ctx, host, path)
-	baseURL := fmt.Sprintf("https://%s/v2/%s", host, path)
+	scheme := "https"
+	if strings.HasSuffix(host, ":5000") { // don't support https on 5000 for now
```
Comment (codefromthecrypt) [inline internal/registry/registry.go:54]: note: I owe some tests like here, but will wait for a review before doing more.

File: internal/registry/registry.go [hunk 2]
```diff
@@ -71,37 +76,28 @@ func (r *registry) String() string {
...
+	p := pathutil.Join(config.OS, config.Architecture)
+	if platform != "" && p != platform {
+		return nil, fmt.Errorf("tag %s is for platform %s, not %s", tag, p, platform)
 	}
-	if len(result) == 0 {
-		return nil, fmt.Errorf("image tag %s not found for platform %s", tag, platform)
-	}
-	return result[len(result)-1], nil
+
+	// Combine the two sources into the Image internal we need.
```
Comment (codefromthecrypt) [inline internal/registry/registry.go:96]:
```suggestion
	// Combine the two sources into the Image we need.
```

File: internal/cmd/app.go [hunk 3]
```diff
@@ -81,11 +82,14 @@ func newApp() *cli.App {
 			} else {
 				return &validationError{fmt.Sprintf("invalid [%s] flag: expected tagged reference", flagReference)}
 			}
-			return validatePlatformFlag(c.String(flagPlatform))
+			if platform = c.String(flagPlatform); platform != "" {
```
Comment (mathetake) [inline internal/cmd/app.go:85]:
```suggestion
			if c.String(flagPlatform) != "" {
```

Comment (mathetake) [reply internal/cmd/app.go:85]: oh, the variable is in outer scope! nvm

File: internal/registry/registry.go [hunk 4]
```diff
@@ -123,52 +119,94 @@ func (r *registry) getImageManifests(ctx context.Context, tag, platform string)
...
 }
 
-func (r *registry) getImageConfigs(ctx context.Context, images []*imageManifestV1) ([]*imageConfigV1, error) {
-	var configs = make([]*imageConfigV1, len(images))
-	for i, image := range images {
-		if !strings.Contains(acceptImageConfigV1, image.Config.MediaType) {
-			return nil, fmt.Errorf("invalid config media type in image %v", image)
+func requireValidPlatform(tag, platform string, platforms map[string]string) (string, error) {
+	// create a list of the platforms we found
+	var s []string //nolint:prealloc
```
Comment (mathetake) [inline internal/registry/registry.go:170]:
```suggestion
	s := make([]string, 0, len(platforms))
```

File: internal/registry/registry.go [hunk 5]
```diff
@@ -123,52 +119,94 @@ func (r *registry) getImageManifests(ctx context.Context, tag, platform string)
...
+	for p := range platforms {
+		s = append(s, p)
+	}
+	sort.Slice(s, func(i, j int) bool {
+		return s[i] < s[j]
+	})
+
+	if platform == "" {
+		switch len(platforms) {
+		case 0:
```
Comment (mathetake) [inline internal/registry/registry.go:180]: I think we can return `return "", fmt.Errorf("tag %s has no platform information", tag)` here rather than check len(platforms) == 0 below.

File: internal/registry/registry.go [hunk 6]
```diff
@@ -123,52 +119,94 @@ func (r *registry) getImageManifests(ctx context.Context, tag, platform string)
...
+	manifest.URL = url
+	return &manifest, nil
 }
 
-func (r *registry) getImageConfigs(ctx context.Context, images []*imageManifestV1) ([]*imageConfigV1, error) {
-	var configs = make([]*imageConfigV1, len(images))
-	for i, image := range images {
-		if !strings.Contains(acceptImageConfigV1, image.Config.MediaType) {
-			return nil, fmt.Errorf("invalid config media type in image %v", image)
+func requireValidPlatform(tag, platform string, platforms map[string]string) (string, error) {
```
Comment (mathetake) [inline internal/registry/registry.go:168]: we can early return if len(platforms) ==0 here!

Comment (codefromthecrypt) [reply internal/registry/registry.go:168]: oo I like early return

**Comment by codefromthecrypt** [comment]
ok I backfilled tests and tightened up error handling based on some edge cases in what the schema allows (notably index where manifests don't have os/arch info)

**Comment by codefromthecrypt** [comment]
thanks for the help on this @mathetake I'm going to merge, but if you have any feedback later I will look into it for sure!
